File: integrations/sqlite/typescript/src/recovery-cursor.ts
/** Host-named traversal progress, never execution ownership. */
type Extra = Record<string, any>;
type Transaction = <T>(operation: (extra: Extra) => Promise<T>, readOnly: boolean) => Promise<T>;
type Page = Readonly<{ runIds: readonly string[]; nextAfterRunId: string | null }>;
type Position = { revision: number; afterRunId: string | null };
export class SqliteRecoveryCursor {
  #page: { position: Position; page: Page } | undefined;
  constructor(private readonly transaction: Transaction,
    private readonly list: (afterRunId: string | null, limit: number) => Promise<Page>,
    private readonly name: string, private readonly pageSize = 100) {
    if (typeof name !== 'string' || !name.trim()) throw new TypeError('cursor name required');
    if (!Number.isInteger(pageSize) || pageSize < 1 || pageSize > 1000) throw new TypeError('invalid cursor page size');
  }
  private row(extra: Extra): Position {
    const rows = extra.recoveryCursors === undefined ? {} : extra.recoveryCursors;
    if (rows === null || typeof rows !== 'object' || Array.isArray(rows)) throw new TypeError('invalid recovery cursors');
    const row = Object.hasOwn(rows, this.name) ? rows[this.name] : { revision: 0, afterRunId: null };
    if (row === null || typeof row !== 'object' || Object.keys(row).sort().join() !== 'afterRunId,revision'
      || !Number.isSafeInteger(row.revision) || row.revision < 0 || row.revision >= Number.MAX_SAFE_INTEGER
      || (row.afterRunId !== null && (typeof row.afterRunId !== 'string' || !row.afterRunId))) throw new TypeError('invalid recovery cursor');
    return row;
  }
  async discover(): Promise<readonly string[]> {
    this.#page = undefined;
    const position = await this.transaction(async extra => ({ ...this.row(extra) }), true);
    const page = await this.list(position.afterRunId, this.pageSize);
    this.#page = {position, page};
    return page.runIds;
  }
  async acknowledge(processedIds: readonly string[]): Promise<void> {
    if (this.#page === undefined) throw new Error('cursor page not discovered');
    const { position, page } = this.#page;
    if (processedIds.length > page.runIds.length || processedIds.some((id, i) => id !== page.runIds[i])) throw new TypeError('cursor acknowledgement must be a processed prefix');
    if (processedIds.length === 0 && (page.runIds.length > 0 || position.afterRunId === null)) return;
    const afterRunId = processedIds.length === page.runIds.length && page.nextAfterRunId === null ? null : processedIds[processedIds.length-1]!;
    await this.transaction(async extra => {
      const current = this.row(extra);
      if (current.revision !== position.revision || current.afterRunId !== position.afterRunId) throw new Error('recovery_cursor_conflict');
      if (position.revision + 1 >= Number.MAX_SAFE_INTEGER) throw new Error('invalid recovery cursor revision');
      const rows = Object.assign(Object.create(null), extra.recoveryCursors ?? {});
      rows[this.name] = {revision:position.revision+1, afterRunId};
      extra.recoveryCursors = rows;
    }, false);
    this.#page = undefined;
  }
}

File: integrations/sqlite/python/src/purra_sqlite/recovery_cursor.py
"""Host-named traversal progress, independent of execution ownership."""
from .recovery_schedule import _integer


class SqliteRecoveryCursor:
    def __init__(self, storage, name, *, page_size=100):
        if not isinstance(name, str) or not name.strip(): raise ValueError('cursor name required')
        if type(page_size) is not int or not 1 <= page_size <= 1000: raise ValueError('invalid cursor page size')
        self._storage, self._name, self._size = storage, name, page_size
        self._page = None

    def _row(self, extra):
        rows = extra.get('recoveryCursors', {})
        if not isinstance(rows, dict): raise ValueError('invalid recovery cursors')
        row = rows.get(self._name, {'revision': 0, 'afterRunId': None})
        if not isinstance(row, dict) or set(row) != {'revision', 'afterRunId'}: raise ValueError('invalid recovery cursor')
        _integer(row['revision'])
        if row['afterRunId'] is not None and (not isinstance(row['afterRunId'], str) or not row['afterRunId']): raise ValueError('invalid recovery cursor position')
        return row

    async def discover(self):
        self._page = None
        async with self._storage._metadata_transaction(read_only=True) as session:
            row = dict(self._row(session.extra))
        page = await self._storage.list_run_candidates(after_run_id=row['afterRunId'], limit=self._size)
        self._page = (row, page)
        return page['runIds']

    async def acknowledge(self, processed_ids):
        if self._page is None: raise ValueError('cursor page not discovered')
        row, page = self._page
        processed = tuple(processed_ids)
        if len(processed) > len(page['runIds']) or processed != page['runIds'][:len(processed)]:
            raise ValueError('cursor acknowledgement must be a processed prefix')
        if not processed and (page['runIds'] or row['afterRunId'] is None): return
        after = (None if len(processed) == len(page['runIds']) and page['nextAfterRunId'] is None
                 else processed[-1])
        async with self._storage._metadata_transaction() as session:
            if self._row(session.extra) != row: raise ValueError('recovery_cursor_conflict')
            session.extra.setdefault('recoveryCursors', {})[self._name] = {
                'revision': _integer(row['revision'] + 1), 'afterRunId': after,
            }
        self._page = None
